common/persistence/cassandra/metadata_store.go

```diff
@@ -161,23 +168,48 @@ func (m *MetadataStore) CreateNamespaceInV2Table(
 
 	previous := make(map[string]interface{})
 	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	deleteOrphanNamespace := func() {
+		// Delete namespace from `namespaces_by_id`
+		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
+			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+		}
+	}
+
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("CreateNamespace operation failed. Inserting into namespaces table. Error: %v", err))
 	}
+
 	defer func() { _ = iter.Close() }()
 
 	if !applied {
-		// Namespace already exist.  Delete orphan namespace record before returning back to user
-		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
-			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+
+		// if both conditions fail, find the one related to the first query
```

Comment:
looks like we need something similar to `covertErrors` & `sortErrors` (under common/persistence/cassandra/errors.go) here. but yeah this impl works as well.

Reply:
I don't think we need something that generic, but we should encapsulate this logic

```go
func fieldMatches[T](row Row, column string, value T) (bool, error) {
  existingValue, ok := row[column]
  if !ok {
    return false, fmt.Errorf("%w: %q", ErrColumnNotFound, column)
  }
  return existingValue == value, nil
}
```

Reply:
@yycptt  nice, `MapExecuteBatchCAS` [removes](https://github.com/gocql/gocql/blob/master/session.go#L771) the `[applied]` field. I added a line to explicitly check the value of `[applied]` from the second row, just to make sure we are not matching against an actual failure.